### PR TITLE
golang 1.16 rpm spec, patches, and project structure

### PR DESCRIPTION
**Pull request by danbudris**
*Issue #, if available:*
https://github.com/aws/eks-distro-internal/issues/221
https://github.com/aws/eks-distro-internal/issues/207

*Description of changes:*
Add patches for go 1.16
Add RPM spec for go 1.16
Add folder structure for project overall

This has been tested by running the rpm build make target on an AL machine; I'm going to hold it for now untill we can run the pre-submits against this change (see https://github.com/aws/eks-distro-prow-jobs/pull/354) first and ensure it builds.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by danbudris** [comment]
pre-submits passing; removed hold

File: projects/golang/go/Makefile [hunk 1]
```diff
@@ -64,7 +64,7 @@ clean:
 
 .PHONY: install-deps
 install-deps:
-	yum update -y && yum install -y which git rpmdevtools awscli rpm-build golang-1.15.14-1.amzn2.0.1 pcre-devel glibc-static hostname procps-ng
+	yum update -y && yum install -y which git rpmdevtools go-srpm-macros awscli rpm-build golang-1.15.14-1.amzn2.0.1 pcre-devel glibc-static hostname procps-ng
```
Comment (jaxesn) [inline projects/golang/go/Makefile:67]: i think for the 1.16 build we should install the 1.16 yum package? or from upstream if it doesn't exist?

File: projects/golang/go/1.16/patches/0101-syscall-expose-IfInfomsg.X__ifi_pad-on-s390x.patch [hunk 2]
```diff
@@ -0,0 +1,41 @@
+From 2e472ee2cfd536a2f62c2c66be47dc831f524beb Mon Sep 17 00:00:00 2001
```
Comment (markapruett) [inline projects/golang/go/1.16/patches/0101-syscall-expose-IfInfomsg.X__ifi_pad-on-s390x.patch:1]: Remind me why this one has a different structure than the other patches?

File: projects/golang/go/1.16/rpmbuild/SOURCES/golang-gdbinit [hunk 3]
```diff
@@ -0,0 +1 @@
+add-auto-load-safe-path /usr/lib/golang/src/runtime/runtime-gdb.py
```
Comment (markapruett) [inline projects/golang/go/1.16/rpmbuild/SOURCES/golang-gdbinit:1]: What does this do?

Comment (danbudris) [reply projects/golang/go/1.16/rpmbuild/SOURCES/golang-gdbinit:1]: gdb is the GNU debugger; Golang includes tests for GDB compatibility. 

In GDB, you can load helper scripts which make it easier to convert your data structures to debugger calls and do other helpful things; in order to include these external helper scripts in the tests, you need to explicitly set their path and allow them to be imported safely.

This gdbinit file does that, and was included with the upstream RPM as a source; we're just carrying it over here

File: projects/golang/go/1.16/rpmbuild/SPECS/golang.spec [hunk 4]
```diff
@@ -0,0 +1,1333 @@
...
+
+%global go_api 1.16
+%global go_version %{go_api}.15
+
+# For rpmdev-bumpspec and releng automation
+%global baserelease 1
+
+Name:           golang
+Version:        %{go_version}
+Release: %{baserelease}%{?dist}.0.1
```
Comment (markapruett) [inline projects/golang/go/1.16/rpmbuild/SPECS/golang.spec:116]: Do we want to have this hardcoded here?

File: projects/golang/go/1.16/patches/0101-syscall-expose-IfInfomsg.X__ifi_pad-on-s390x.patch (hunk 2 above)
Comment (danbudris) [reply projects/golang/go/1.16/patches/0101-syscall-expose-IfInfomsg.X__ifi_pad-on-s390x.patch:1]: yeah these are patches that were included by upstream in the RPM; they have a different format than our patches, which I prefaced with the version branch we used as the source. happy to change the format on these guys if you have suggestions

Comment (markapruett) [reply projects/golang/go/1.16/patches/0101-syscall-expose-IfInfomsg.X__ifi_pad-on-s390x.patch:1]: Fine with me for now until we learn otherwise.

**Review by markapruett — APPROVED** [review]

File: projects/golang/go/Makefile (hunk 1 above)
Comment (danbudris) [reply projects/golang/go/Makefile:67]: will address this in a following PR

**Pull request merged**